.github/workflows/reuse.yml

```diff
@@ -0,0 +1,254 @@
+        # SECUREBOOT: ${{ inputs.SECUREBOOT }}
+      run: |
+        make config
+        make build
+        docker image ls -a
+
+    - name: Save the images
+      shell: bash
+      run: |
+        docker save --output /tmp/corelatest.tar leviathan_core:latest 
+        docker save --output /tmp/clientlatest.tar leviathan_client:latest 
+
+    - name: Upload Core artifact
+      uses: actions/upload-artifact@v3
+      with:
+        name: corelatest.tar
+        path: /tmp/corelatest.tar
+        compression-level: 7
+
+
+    - name: Upload Client artifact
+      uses: actions/upload-artifact@v3
+      with:
+        name: clientlatest.tar
+        path: /tmp/clientlatest.tar
+        compression-level: 7
+  
+
+  test:
+    runs-on: ubuntu-22.04
+    needs: build
+    steps:
+    - name: Download artifact
+      uses: actions/download-artifact@v3
+      with:
+        name: clientlatest.tar
+        path: /tmp
+
+    - name: Download artifact
+      uses: actions/download-artifact@v3
+      with:
+        name: corelatest.tar
+        path: /tmp
+
+    - name: Load image
+      shell: bash
+      run: |
+        docker image ls -a
+        docker load --input /tmp/corelatest.tar
+        docker load --input /tmp/clientlatest.tar
+        docker image ls -a
+
+    - name: Fetching code to be tested
+      uses: actions/checkout@v4
+      with:
+        submodules: true
+
+    - name: Get Leviathan checked out
+      uses: actions/checkout@v4
+      with:
+        repository: balena-os/leviathan
+        submodules: true
+        path: ./leviathan
+        # The branch, tag or SHA to checkout. When checking out the repository that
+        # triggered a workflow, this defaults to the reference or SHA for that event.
+        # Otherwise, uses the default branch.
+        ref: ${{ inputs.LEVIATHAN_REF }}
+
+
+    # - name: Fetching the Private Contracts
+    #   uses: actions/checkout@v4
+    #   with:
+    #     repository: balena-io/private-contracts
+    #     path: './leviathan/core/private-contracts'
+
+    - name: All the copying stage
+      shell: bash
+      working-directory: ./leviathan
+      run: |
+        mkdir -p ${{ inputs.WORKSPACE }}
+        (cd ${{ inputs.WORKSPACE }} && mkdir -p ${{ inputs.REPORTS }})
+        cp ../${{ inputs.SUITES }}/config.js ${{ inputs.WORKSPACE }}
+        cp -r ../${{ inputs.SUITES }} .
+
+    # - name: Show the final config.js before RUN
+    #   working-directory: ./leviathan-workspace
+    #   run: |
+    #     cat ${{ inputs.WORKSPACE }}/config.js
+    #   shell: bash
+
+    - name: generate a unique subnet for this job
+      working-directory: ./leviathan
+      run: |
+        make subnet || true
+        make config
+      shell: bash
+
+    - name: Test
+      working-directory: ./leviathan
+      shell: bash
+      env:
+        BALENACLOUD_API_KEY: ${{ secrets.BALENACLOUD_API_KEY }}
+        BALENAMACHINE_API_KEY: ${{ secrets.BALENAMACHINE_API_KEY }}
```

Comment:
I think Kyle has already said this basically, but there should just be one API_KEY//URL/SSH_PORT .. etc, and the selection being made as either a step seomwhere else, or via environments . That way we can add other envs like staging more sustainably without expanding this list

Reply:
They should be read from the GitHub Environment `vars` value for workflows, but in a composite action they should be inputs passed in by the workflow.